### ISSUE TITLE
[Bug]: Physics check ignores the GGUF's own sliding_window_pattern — unknown SWA architectures are priced as fully global and refused (gemma4: 32.1 GiB estimated vs ~11 GiB actual)

### ISSUE THREAD

**Issue by wyuebei-cloud** [NousResearch/hermes-agent#109551]
### Bug Description

`hermes_cli/local_runtime/estimator.py` decides how much memory a model needs. For
sliding-window-attention (SWA) architectures it does **not** read the per-layer pattern that the
GGUF file already declares (`{arch}.attention.sliding_window_pattern`). It consults a hardcoded
architecture-name table instead:

```python
# hermes_cli/local_runtime/estimator.py:20
_SWA_LAYER_FRACTION = {"gemma3": 5 / 6, "gemma2": 1 / 2}
```

Any architecture not in that table falls through to "price every layer as full attention". On a
48-layer `gemma4` model where 40 of the 48 layers are sliding-window (window = 1024), this inflates
the KV cache at 64K context from **0.70 GiB to 21.78 GiB**, and the whole footprint from
**11.06 GiB to 32.1 GiB**. The physics check then refuses a model that fits comfortably and runs
fine on the same machine with the same binary.

The failure is silent: the model never appears in `GET /v1/models`, and a completion returns
`400 model 'X' not found`. Nothing surfaces the fact that the estimate — not the hardware — is what
rejected it.

I am filing this as a design problem, not as "please add `gemma4` to the table". The table is a
losing race: the information is already in the file, per layer, for every model that follows the
GGUF convention, and a name table has to be extended by a Hermes release every time a new SWA
architecture ships. See **Root Cause Analysis** and **Proposed Fix**.

### Steps to Reproduce

1. Take any GGUF whose architecture declares a per-layer SWA pattern but is absent from
   `_SWA_LAYER_FRACTION`. Here: `general.architecture = gemma4`
   (`gemma4-v2-Q4_K_M.gguf`, hand-downloaded from Hugging Face, not a catalog entry).
2. Place it in `$HERMES_HOME/models/`.
3. Boot the managed local runtime so `generate_presets()` runs.

### Expected Behavior

The model should be priced from the attention layout its own file declares, pass the physics check,
and be servable: `GET /v1/models` lists `gemma4-v2-Q4_K_M`, and a completion returns text.

More generally: an architecture that is absent from a hardcoded table should not, by itself, be
grounds for refusal when the file declares exactly the information the estimator needs to compute
the footprint correctly.

### Actual Behavior

```
WARNING hermes_cli.local_runtime.bootstrap: model refused by physics check:
  <models>/gemma4-v2-Q4_K_M.gguf: needs ~32.1 GiB at the 64K floor
  but only ~23.1 GiB of VRAM+RAM are available — try a smaller model or a supported smaller quant

INFO hermes_cli.local_runtime.presets: wrote 0 preset sections to <runtimes>/llamacpp/presets.ini
```

`presets.ini` ends up containing only the `# hermes-decision:` refusal record and **no model
section**, so the router has nothing to serve:

```
GET /v1/models
  {"data":[]}

POST /v1/chat/completions  {"model": "gemma4-v2-Q4_K_M", ...}
  {"error":{"code":400,"message":"model 'gemma4-v2-Q4_K_M' not found","type":"invalid_request_error"}}
```

Estimated vs. correct footprint for the same file:

| | As shipped | Correct, from the file's per-layer metadata |
|---|---|---|
| per-token KV (f16, uncapped sum) | 656 KB | per-layer, see below |
| KV @ 64K | 21.78 GiB | **0.70 GiB** |
| weights | 6.86 GiB | 6.86 GiB |
| fixed overhead (runtime 1.5 + logits buffer 2.0) | 3.50 GiB | 3.50 GiB |
| **total @ 64K** | **32.1 GiB → refused** | **11.06 GiB → fits** |
| total @ 256K (native window) | 87 GiB | 12.65 GiB |

Budget on this machine: 23.07 GiB (`uma=True`, 28.83 GiB RAM × 0.80 UMA headroom).

The model does run. Same binary, same file, started directly without a Hermes preset:

```
llama-server.exe --model <models>/gemma4-v2-Q4_K_M.gguf \
  --host 127.0.0.1 --port 18436 --ctx-size 65536 \
  --flash-attn on --cache-type-k q8_0 --cache-type-v q8_0

0.08.568 I srv load_model: model loaded          (8.6 s from spawn)
```

A chat completion returns coherent text, resident working set **10.6 GiB**. Served through the
Hermes router afterwards, it loaded at `n_ctx_slot = 221184` — the full window the launch policy
wanted — and stayed at ~11.8 GiB.

For reference, a one-line table patch (`{"gemma4": 5 / 6, ...}`) flips the decision from
`window=0 (refused)` to `window=221184 (accepted)`. That is evidence that this single table lookup
is the whole gate — and it is also only a coarse approximation, which is the point of the fix below.

### Affected Component

Setup / Installation — the managed local runtime (`hermes_cli/local_runtime/`), model launch-policy
generation.

### Messaging Platform

N/A (CLI only)

### Debug Report

Not uploaded. `hermes debug share` uploads config plus recent logs, which for this report would
include unrelated session content well beyond the scope of the bug. The information the bundle
would contribute is provided inline instead:

```
Hermes Agent   v0.21.2 (2026.9.11) · upstream 1c671bea, install method: git
Python         3.11.16
OS             Windows 11 x64
Hardware       AMD Radeon 780M iGPU (no discrete GPU), 30.96 GB RAM
Budget         uma=True, 23.07 GiB
llama.cpp      tag b10679, cpu backend
Model          gemma4-v2-Q4_K_M.gguf, 7,381,381,664 bytes,
               general.architecture = gemma4, 48 layers, native ctx 262144
```

Happy to run `hermes debug share` and paste the links if you want the full bundle.

### Operating System

Windows 11 x64

### Python Version

3.11.16

### Hermes Version

Hermes Agent v0.21.2 (2026.9.11) · upstream `1c671bea`, install method: git

### Additional Logs / Traceback (optional)

The complete trace of the refusal, including the decision record written next to the preset file:

```
# <runtimes>/llamacpp/presets.ini
# hermes-decision: {"model_id": "gemma4-v2-Q4_K_M", "window": 0, "spilled": false,
#                   "refusal": "<models>/gemma4-v2-Q4_K_M.gguf: needs ~32.1 GiB at the 64K floor
#                   but only ~23.1 GiB of VRAM+RAM are available — try a smaller model or a
#                   supported smaller quant"}
```

No traceback: the refusal is a normal, logged decision, not an exception.

### Root Cause Analysis (optional)

**What the file declares.** Read straight out of the GGUF header:

```
gemma4.block_count                        = 48
gemma4.context_length                     = 262144
gemma4.attention.head_count_kv            = [8,8,8,8,8,1] x 8
gemma4.attention.sliding_window           = 1024
gemma4.attention.sliding_window_pattern   = [1,1,1,1,1,0] x 8     # 40 SWA / 8 global
gemma4.attention.key_length               = 512
gemma4.attention.key_length_swa           = 256
gemma4.attention.value_length             = 512
gemma4.attention.value_length_swa         = 256
```

`sliding_window_pattern` is an explicit 48-entry array: `1` marks a sliding-window layer, `0` a
global one. The file is not silent about its own attention layout.

**What the estimator reads.**

```python
# estimator.py:79-85
swa_fraction = _SWA_LAYER_FRACTION.get(header.architecture, 0.0)   # miss -> 0.0
has_swa = header.sliding_window > 0 and swa_fraction > 0           # -> False
# n_swa = 0, so every attention layer is appended as LayerKind.FULL
```

`gguf.py` has accessors for `attention.sliding_window`, `attention.key_length` and
`attention.value_length`. It has **no accessor at all** for `attention.sliding_window_pattern`,
`attention.key_length_swa` or `attention.value_length_swa`.

Secondary defect in the same area: `head_dim_k` / `head_dim_v` (`gguf.py:137-149`) resolve to
`attention.key_length` / `attention.value_length` (512) and are then applied to **every** layer.
The SWA layers declare 256. So even for architectures already present in the table (e.g. `gemma3`),
the SWA share of the KV is priced at roughly 2x its real per-token cost.

**The inconsistency worth fixing.** The estimator already trusts per-layer metadata for the other
context-growth mechanism:

```python
# gguf.py:119-135 — head_counts_kv()
#   a per-layer array is used as-is; head_count_kv == 0 marks a recurrent/linear layer
# gguf.py:78-81 — full_attention_interval
#   "GDN-hybrid discriminator (qwen35 family): every Nth layer is full attention"
```

and `estimator.profile_from_gguf` prices those layers as a constant
(`_RECURRENT_STATE_PER_LAYER`) rather than as growing KV. That path exists precisely because a name
table could not answer it.

So: per-layer metadata is trusted for **recurrent** layers, and a hardcoded name table is used for
**sliding-window** layers — even though the SWA pattern is declared just as explicitly, in the same
header, for the same reason.

**Why "overestimate; safe" is not safe here.** The comment above the table states the current
behaviour is intentional:

```python
# Unknown SWA archs treat every layer as full attention (overestimate; safe).
```

It is safe for the engine and unsafe for the user. The outcome is a **false negative**: a model
that runs is refused, and the refusal is indistinguishable from "your machine really can't do
this". There is no log line, badge, or hint that the estimate is the reason — the user sees an
empty model list. Here it rejected a model that occupies roughly one third of the budget.

The blast radius is also skewed. Hand-downloaded GGUFs — the case a user cannot work around without
editing Hermes source — take the unknown-architecture path. Extending the table per architecture
only moves where the next false negative lands.

### Proposed Fix (optional)

Make the file's own metadata authoritative, and keep the architecture table as a fallback. Priority
ladder, highest first:

1. **Per-layer array, when present** — read `{arch}.attention.sliding_window_pattern` and build the
   layer list from it directly; use `{arch}.attention.key_length_swa` /
   `{arch}.attention.value_length_swa` for the sliding-window layers. Architecture-agnostic; no
   Hermes change when a new SWA family ships.
2. **Window + known-architecture fraction** — current behaviour, for files that declare
   `sliding_window` but no per-layer pattern (older GGUFs).
3. **All layers global** — current worst-case fallback, only when neither of the above exists.

Step 1 also fixes the secondary defect: it gives the SWA layers their own dimensions instead of
reusing the global ones.

If a coarser change is preferred in the short term, adding `gemma4` to `_SWA_LAYER_FRACTION` does
unblock this file, but it leaves the `*_swa` dimensions wrong (0.70 GiB vs 3.61 GiB of KV at 64K —
still conservative, still fitting) and does not stop the next architecture from hitting the same
wall.

Scope: not `gemma4`-specific. Any architecture that declares a per-layer SWA pattern and is not
enumerated in `_SWA_LAYER_FRACTION` is priced as fully global. The Gemma family is simply the common
case (5 of every 6 layers sliding-window), which is why the error lands near 3x rather than at a few
percent.

### Are you willing to submit a PR for this?

No — reporting only.
